packages/query-core/src/queryClient.ts

```diff
@@ -366,92 +366,52 @@ export class QueryClient {
-        partialMatchKey(queryKey, x.queryKey),
-      )
-      // It is ok not having defaults, but it is error prone to have more than 1 default for a given key
-      if (matchingDefaults.length > 1) {
-        this.#logger.error(
-          `[QueryClient] Several query defaults match with key '${JSON.stringify(
-            queryKey,
-          )}'. The first matching query defaults are used. Please check how query defaults are registered. Order does matter here. cf. https://react-query.tanstack.com/reference/QueryClient#queryclientsetquerydefaults.`,
-        )
-      }
-    }
-
-    return firstMatchingDefaults?.defaultOptions
+    return matchingDefaults.reduce((accumulator, matchingDefault) => {
+      return { ...accumulator, ...matchingDefault.defaultOptions }
+    }, {})
   }
 
   setMutationDefaults(
     mutationKey: MutationKey,
-    options: MutationObserverOptions<any, any, any, any>,
+    options: Omit<MutationObserverOptions<any, any, any, any>, 'mutationKey'>,
   ): void {
-    const result = this.#mutationDefaults.find(
-      (x) => hashKey(mutationKey) === hashKey(x.mutationKey),
-    )
-    if (result) {
-      result.defaultOptions = options
-    } else {
-      this.#mutationDefaults.push({ mutationKey, defaultOptions: options })
-    }
+    this.#mutationDefaults.set(hashKey(mutationKey), {
+      mutationKey,
+      defaultOptions: options,
+    })
   }
 
   getMutationDefaults(
-    mutationKey?: MutationKey,
+    mutationKey: MutationKey,
   ): MutationObserverOptions<any, any, any, any> | undefined {
-    if (!mutationKey) {
-      return undefined
-    }
+    const defaults = [...this.#mutationDefaults.values()]
 
-    // Get the first matching defaults
-    const firstMatchingDefaults = this.#mutationDefaults.find((x) =>
+    const matchingDefaults = defaults.filter((x) =>
       partialMatchKey(mutationKey, x.mutationKey),
     )
 
-    // Additional checks and error in dev mode
-    if (process.env.NODE_ENV !== 'production') {
-      // Retrieve all matching defaults for the given key
-      const matchingDefaults = this.#mutationDefaults.filter((x) =>
-        partialMatchKey(mutationKey, x.mutationKey),
-      )
-      // It is ok not having defaults, but it is error prone to have more than 1 default for a given key
-      if (matchingDefaults.length > 1) {
-        this.#logger.error(
-          `[QueryClient] Several mutation defaults match with key '${JSON.stringify(
-            mutationKey,
-          )}'. The first matching mutation defaults are used. Please check how mutation defaults are registered. Order does matter here. cf. https://react-query.tanstack.com/reference/QueryClient#queryclientsetmutationdefaults.`,
-        )
-      }
-    }
-
-    return firstMatchingDefaults?.defaultOptions
+    return matchingDefaults.reduce((accumulator, matchingDefault) => {
+      return { ...accumulator, ...matchingDefault.defaultOptions }
+    }, {})
```

Comment:
An alternate version, but your's is fine, this one evoids allocating an intermediate array :
```ts
  let result:  MutationObserverOptions<any, any, any, any> = {}
  this.#mutationDefaults.forEach( (default) => {
    if (partialMatchKey(mutationKey, default.mutationKey)) {
      result = { ...result, default.defaultOptions }
    }
  });
  return result
```